### ISSUE TITLE
Autoconfiguration for GPU way off

### ISSUE THREAD

**Issue by sheepchen** [fireice-uk/xmr-stak#88]
Hey,
I compiled the xmr-stak version for windows and tested it.

My mainworkingstation (Testsystem):
Windows 10 Pro 64 Bit
Intel i7-4790K 4Ghz uc@3.6 Ghz
Nvidia Geforce GTX 980 4GB

configs (generated automaticly):
```
"cpu_threads_conf" :
[
    { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 0 },
    { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 2 },
    { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 4 },
    { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 6 },

],
```
```
"gpu_threads_conf" :
[
  // gpu: GeForce GTX 980 architecture: 52
  //      memory: 3376/4096 MiB
  { "index" : 0,
    "threads" : 20, "blocks" : 48,
    "bfactor" : 6, "bsleep" :  25,
    "affine_to_cpu" : false,
  },

],
```
First run crashed directly:
![grafik](https://user-images.githubusercontent.com/32341444/32369184-9935307e-c088-11e7-953c-a774e041181e.png)

I changed nvidia config like this:
bfactor: 8
bsleep: 100

result: program runs but the System was very laggy.

I changed nvidia config again like this:
bfactor: 12
bsleep: 100

result: program runs and system is only a little bit laggy (an vlc-video runs after 2 sec smootly)

after 15 minutes:
```
HASHRATE REPORT
| ID | 10s |  60s |  15m | ID | 10s |  60s |  15m |
|  0 | (na) | 84.3 | 84.8 |  1 | 31.1 | 31.5 | 32.9 |
|  2 | 38.9 | 37.2 | 37.9 |  3 | 38.2 | 36.3 | 36.1 |
|  4 | 21.8 | 22.0 | 22.9 |
-----------------------------------------------------
Totals:   (na) 211.2 214.5 H/s
Highest:  0.0 H/s
```

So i tweaked the config a little bit:
```
"cpu_threads_conf" : 
[ 
     { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 2 },
     { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 3 },
     { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 4 },
     { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 5 },
     { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 6 },
     { "low_power_mode" : false, "no_prefetch" : true, "affine_to_cpu" : 7 },
],
```
// 2,4,6 are real cores
// 3,5,7 are HT cores

```
 "gpu_threads_conf" : 
 [
     { "index" : 0, "threads" : 32, "blocks" : 16, "bfactor" : 12, "bsleep" : 5, "affine_to_cpu" : true},
 ],
```

result:
```
HASHRATE REPORT
| ID | 10s |  60s |  15m | ID | 10s |  60s |  15m |
|  0 | 467.4 | 461.6 | 459.6 |  1 | 20.4 | 19.4 | 19.1 |
|  2 | 25.2 | 25.2 | 25.8 |  3 | 12.5 | 15.2 | 13.9 |
|  4 | 23.1 | 22.4 | 23.0 |  5 | 16.3 | 16.3 | 17.1 |
|  6 | 33.5 | 30.3 | 31.3 |
-----------------------------------------------------
Totals:   598.3 590.4 589.7 H/s
Highest:  642.1 H/s
```

here is how i find the best values of my cards:
[xmr-stat-gpu values.xlsx](https://github.com/fireice-uk/xmr-stak/files/1440774/xmr-stat-gpu.values.xlsx)

Conclusion: 
 - HT cores are not effective (4 cores --> 129,8 H/s     3 cores + 3 HT --> 130,2 H/s)
   --> the automatic values for CPU seems okay.
 - the automatically determined values for GPU lead to a crash
 - even for-crash corrected values are scrap (84 H/s)
 - better test your GPU by hand. Tool for GPU-MEM-check https://www.techpowerup.com/gpuz/

**Comment by psychocrypt** [comment]
Could you please test the suggested gpu config and change threads to 16.
It is strange that the auto suggestion is crashing. Could you please also
post the cmake command.

**Comment by sheepchen** [comment]
1. I build the system straight the commands descriped in comple_Windows.md . No special commands.
2. This is the successed config with your change:
```
"gpu_threads_conf" :
[
  // gpu: GeForce GTX 980 architecture: 52
  //      memory: 3376/4096 MiB
  { "index" : 0,
    "threads" : 16, "blocks" : 48,
    "bfactor" : 6, "bsleep" :  25,
    "affine_to_cpu" : false,
  },
],
```

```
HASHRATE REPORT
| ID | 10s |  60s |  15m | ID | 10s |  60s |  15m |
|  0 | 115.7 | 115.5 | (na) |  1 | 36.6 | 28.4 | (na) |
|  2 | 48.3 | 48.1 | (na) |  3 | 40.9 | 32.8 | (na) |
|  4 | 36.1 | 31.5 | (na) |
-----------------------------------------------------
Totals:   277.6 256.4 (na) H/s
Highest:  0.0 H/s
```

The binary didn't crash with this config, but the system is really sluggish. Its not fun to work with it. Even the mousecurser lags 4-5 secounds behind. Also I only get 115 H/s and not 450 H/s that I get with my nice config (besides even small gaming possible) or 550 H/s when I push the system to the limit (no more working possible).

**Comment by psychocrypt** [comment]
Could you please also test 19 threads and 18 threads.
I found one point which is wrong in the auto suggestion and you background
running processes can be the reason for the bad performance. I know how I
can build a better auto suggestion but need the two resulta to be sure that
I do the right optimization.

**Comment by sheepchen** [comment]
Not today.. tommorow I will test this. Until tomorrow you can write me more testscenarios I shall test.

**Comment by psychocrypt** [comment]
@sheepchen Thx for testing: Could you please also test the #89 (please let the miner create a new auto suggestion)

You can download a zip file with the fix from https://github.com/psychocrypt/xmr-stak/archive/topic-optimizeNVIDIAAutosuggestion.zip

**Comment by sheepchen** [comment]
here are the results
--------------------
xmr-stak-dev_20171001:
threads 20: crash
threads 19: crash
threads 18: crash
threads 17: 5 seconds extremely slow - then crash
threads 16: all extremly slow, 115 H/s, 1913/4096MB GPUMEM used

xmr-stak-topic-optimizeNVIDIAAutosuggestion_dev_20171104:
[buildprocess.log](https://github.com/fireice-uk/xmr-stak/files/1442805/buildprocess.log)
default config:
```
"gpu_threads_conf" :
[
  // gpu: GeForce GTX 980 architecture: 52
  //      memory: 3376/4096 MiB
  { "index" : 0,
    "threads" : 12, "blocks" : 48,
    "bfactor" : 6, "bsleep" :  25,
    "affine_to_cpu" : false,
  },
],
```
result:
```
HASHRATE REPORT
| ID | 10s |  60s |  15m | ID | 10s |  60s |  15m |
|  0 | 392.4 | 385.7 | (na) |  1 | 31.2 | 31.7 | (na) |
|  2 | 42.6 | 41.6 | (na) |  3 | 37.4 | 37.5 | (na) |
|  4 | 39.4 | 37.2 | (na) |
-----------------------------------------------------
Totals:   542.9 533.8 (na) H/s
Highest:  543.8 H/s
```
way better results since last build but the system is extremly laggy (to the point of not usable).

a little bit of tweaking:
bfactor: 10
```
HASHRATE REPORT
| ID | 10s |  60s |  15m | ID | 10s |  60s |  15m |
|  0 | 386.4 | 384.9 | (na) |  1 | 29.1 | 33.4 | (na) |
|  2 | 32.0 | 37.6 | (na) |  3 | 30.8 | 34.7 | (na) |
|  4 | 29.1 | 32.7 | (na) |
-----------------------------------------------------
Totals:   507.5 523.3 (na) H/s
Highest:  536.6 H/s
```
result: almost same results but the system is really smooth to work with.

**Comment by psychocrypt** [comment]
Solved with #89

Note: We can not change the `bfactor` because you case is a special case where you will mine during you work with the system. Many miner needs a fast miner and interactivity is not needed.
Thanks for all you tests, maybe I will add a option into the release after the next that the user can choose between interactivity and performance during the first start.
